# HASLR: fast hybrid assembly of long reads

- [Introduction](#introduction)
- [Installation](#installation)
  - [Requirements](#requirements)
  - [Dependencies](#dependencies)
  - [Building from source](#building)
- [Running](#running)
  - [Options](#options)
- [Bug report](#bugs)
- [Copyright and License](#license)
- [Author](#author)

## <a name="about"></a>Introduction
HASLR is a tool for rapid genome assembly of long sequencing reads. HASLR is a hybrid tool which means it requires long reads generated by Third Generation Sequencing technologies (such as PacBio or Oxford Nanopore) together with Next Generation Sequencing reads (such as Illumina) from the same sample. HASLR is capable of assembling large genomes on a single computing node. Our experiments show that it can assemble a CHM1 human dataset in less than 10 hours using 64 CPU threads.

## <a name="installation"></a>Installation

### <a name="requirements"></a>Requirements
- GCC â‰¥ 4.8.5
- Python3
- zlib

### <a name="dependencies"></a>Dependencies
HASLR depends on the following tools which will be installed automatically:
- [SPOA](https://github.com/rvaser/spoa) - For consensus calling of long reads
- [Minia](https://github.com/GATB/minia) - For assembling short reads
- [minimap2](https://github.com/lh3/minimap2) - For aligning short read contigs onto long reads
- [fastutils](https://github.com/haghshenas/fastutils) - For FASTA/Q manipulation

### <a name="building"></a>Building from source
```
git clone https://github.com/vpc-ccg/haslr.git
cd haslr
make
```

## <a name="running"></a>Running
```
haslr.py [-t THREADS] -o OUT_DIR -g GENOME_SIZE -l LONG -x LONG_TYPE -s SHORT [SHORT ...]
```

### <a name="options"></a>Options
```
required arguments:
  -o, --out OUT_DIR              output directory
  -g, --genome GENOME_SIZE       estimated genome size; accepted suffixes are k,m,g
  -l, --long LONG                long read file
  -x, --type LONG_TYPE           type of long reads chosen from {pacbio,nanopore}
  -s, --short SHORT [SHORT ...]  short read file

optional arguments:
  -t, --threads THREADS          number of CPU threads to use [1]
  --cov-lr COV_LR                amount of long read coverage to use for assembly [25]
  --aln-block ALN_BLOCK          minimum length of alignment block [500]
  --aln-sim ALN_SIM              minimum alignment similarity [0.85]
  --edge-sup EDGE_SUP            minimum number of long read supporting each edge [3]
  --minia-kmer MINIA_KMER        kmer size used by minia [49]
  --minia-solid MINIA_SOLID      minimum kmer abundance used by minia [3]
  --minia-asm MINIA_ASM          type of minia assembly chosen from {contigs,unitigs} [contigs]
  -v, --version                  print version
  -h, --help                     show this help message and exit
```

## <a name="bugs"></a>Bug report
Please report the bugs through HASLR's issue tracker at [https://github.com/vpc-ccg/haslr/issues](https://github.com/vpc-ccg/haslr/issues).

## <a name="license"></a>Copyright and License
This software is released under GNU General Public License (v3.0)
- SPOA is released under MIT license
- Minia is released under AGPL license 
- minimap2 is released under MIT license
- fastutils is released under GPL license 

## <a name="author"></a>Author
[Ehsan Haghshenas](https://github.com/haghshenas) (ehaghshe AT sfu DOT ca)
